src/core/internal/elf/io.d

```diff
@@ -283,37 +284,65 @@ version (LinuxOrBSD)
     alias ElfSection = NativeElfIO.ElfSection;
 }
 
-private struct MMapRegion(T)
+private struct TypedMMapRegion(T)
 {
 @nogc nothrow:
+
     this(int fd, size_t offset, size_t length = 1)
+    {
+        const pageOffset = (offset / pageSize);
+        const offsetDiff = offset - (pageOffset * pageSize);
+        const mappedSize = (length * T.sizeof) + offsetDiff;
+        const pageMapped = (mappedSize / pageSize) + !!(mappedSize % pageSize);
+        this.region = MMapRegion(fd, pageOffset, pageMapped);
+        this.data = cast(const(T)*) (this.region.data.ptr + offsetDiff);
+        if (this.region.data.ptr !is null)
+            this.data = cast(const(T)*) (this.region.data.ptr + offsetDiff);
+    }
+
+    private MMapRegion region;
+
+    public const(T)* data;
+    alias data this;
+}
+
+private struct MMapRegion
+{
+    @nogc nothrow:
+
+    /**
+     * Instantiate an instance of this struct with the provided offset / length
+     *
+     * Params:
+     *   fd = The file descriptor to map (likely obtained with `open`)
+     *   pageOffset = The number of pages to offset the `fd` with.
+     *                For example, if one wishes to skip the first 16 KiB of
+     *                the file, the pageOffset will be 4 (for pageSize == 4kiB).
+     *   pageCount = The number of pages to map
+     */
+    this(int fd, size_t pageOffset, size_t pageCount)
     {
         if (fd == -1)
             return;
 
-        const pageSize = sysconf(_SC_PAGESIZE);
-        const pagedOffset = (offset / pageSize) * pageSize;
-        const offsetDiff = offset - pagedOffset;
-        const mappedSize = length * T.sizeof + offsetDiff;
-
-        auto ptr = cast(ubyte*) mmap(null, mappedSize, PROT_READ, MAP_PRIVATE, fd, pagedOffset);
+        const offset = pageOffset * pageSize;
+        const mappedSize = pageCount * pageSize;
 
-        mappedRegion = ptr[0 .. mappedSize];
-        data = cast(T*) (ptr + offsetDiff);
+        auto ptr = mmap(null, mappedSize, PROT_READ, MAP_PRIVATE, fd, offset);
+        if (ptr !is MAP_FAILED)
+            this.data = (cast(const(ubyte)*) ptr)[0 .. mappedSize];
```

Comment:
I think you missed one of my suggestions:

```D
        const ptr = cast(ubyte*) mmap(null, mappedSize, PROT_READ, MAP_PRIVATE, fd, offset);
        if (ptr !is MAP_FAILED)
            this.data = ptr[0 .. mappedSize];
 ```

Reply:
I just wasn't so found of it, but suggestion applied nonetheless :)